etl-pipeline/processes/record_candidate_finder.py

```diff
@@ -0,0 +1,194 @@
+        5. Repeat up to MAX_MATCH_DISTANCE times
+        
+        Raises:
+            Exception: If the number of candidates exceeds CANDIDATE_POOL_LIMIT
+        """
+        tokenized_record_title = self._tokenize_title(record.title)
+        ids_to_check = {
+            id for id in record.identifiers if re.search(self.IDENTIFIERS_TO_MATCH, id)
+        }
+
+        candidate_record_ids = set()
+        checked_ids = set()
+
+        for match_distance in range(0, self.MAX_MATCH_DISTANCE):
+            matched_records = self._get_records_matching_identifiers(
+                list(ids_to_check), candidate_record_ids.copy()
+            )
+
+            if not matched_records:
+                break
+
+            checked_ids.update(ids_to_check)
+            ids_to_check.clear()
+
+            for (
+                matched_record_title,
+                matched_record_id,
+                matched_record_identifiers,
+                *_,
+            ) in matched_records:
+                if not matched_record_title:
+                    logger.warning("Invalid title found in matched records")
+                    continue
+
+                tokenized_matched_record_title = self._tokenize_title(
+                    matched_record_title
+                )
+
+                if match_distance > 0 and not self._titles_overlap(
+                    tokenized_record_title, tokenized_matched_record_title
+                ):
+                    continue
+
+                ids_to_check.update(
+                    {
+                        id
+                        for id in matched_record_identifiers
+                        if re.search(self.IDENTIFIERS_TO_MATCH, id)
+                        and id not in checked_ids
+                    }
+                )
+                candidate_record_ids.add(matched_record_id)
+
+        if len(candidate_record_ids) > self.CANDIDATE_POOL_LIMIT:
+            raise Exception(
+                f"Candidate pool size {len(candidate_record_ids)} exceeds limit of {self.CANDIDATE_POOL_LIMIT}"
+            )
+
+        return list(candidate_record_ids)
+    
+    def _get_records_matching_identifiers(
+        self, identifiers: List[str], already_matched_record_ids: List[str]
+    ) -> List[Tuple]:
+        batch_size = 100
+        matched_records = []
+
+        for i in range(0, len(identifiers), batch_size):
+            id_batch = self._format_identifiers(identifiers[i : i + batch_size])
+
+            try:
+                records = (
+                    self.db_manager.session.query(
+                        Record.title, Record.id, Record.identifiers, Record.has_part
+                    )
+                    .filter(~Record.id.in_(already_matched_record_ids))
+                    .filter(Record.identifiers.overlap(id_batch))
+                    .filter(Record.title.isnot(None))
+                    .all()
+                )
+
+                matched_records.extend(records)
```

Comment:
We may want to consider understanding the performance of this query at some point!